### PR TITLE
DMIC and headset MIC LED support on sof-icl-rt711-rt1308-rt715

### PR DESCRIPTION
**Pull request by libinyang**
This patchset add the DMIC and headset MIC LED support on sof-icl-rt711-rt1308-rt715.

### CODE REVIEW

File: tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4 [hunk 1]
```diff
@@ -71,7 +73,7 @@ PIPELINE_PCM_ADD(sof/pipe-volume-playback.m4,
 
 # Low Latency capture pipeline 5 on PCM 4 using max 2 channels of s32le.
 # Schedule 48 frames per 1000us deadline on core 0 with priority 0
-PIPELINE_PCM_ADD(sof/pipe-highpass-capture.m4,
+PIPELINE_PCM_ADD(sof/pipe-eq-capture.m4,
```
Comment (lgirdwood) [inline tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4]: @singalsu can you comment here. @libinyang why is this change needed ? it does not appear in commit message.

Comment (singalsu) [reply tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4]: I don't know why this change is done but pipe-eq-capture adds an IIR high-pass with +20 dB gain and pipe-highpass-capture adds an IIR with 0 dB gain. It's possible to get the same gain by setting volume control to +20 dB with the latter topology. But since volume defaults to 0 dB and apparently we don't have a mechanism to pass a default gain from topology to driver so using the topology with gain offset is justified in some cases (e.g. need out of box Ubuntu to sound sufficiently loud). @juimonen and @libinyang know would be the experts to comment if some UCM setting could achieve the same.

File: tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4 [hunk 2]
```diff
@@ -13,6 +13,8 @@ include(`common/tlv.m4')
 # Include Token library
 include(`sof/tokens.m4')
 
+include(`platform/intel/dmic.m4')
```
Comment (plbossart) [inline tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4]: @libinyang are you aware that we don't use the PCH-attached DMICs on this platform?

All DMICs are connected to the RT715. So is this patch even relevant?

File: tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4 (hunk 1 above)
Comment (plbossart) [inline tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4]: this change does not look correct, we recently had @singalsu add a high-pass filter to remove a DC component on startup, why would we want to modify this?

Comment (paulstelian97) [reply tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4]: I have already seen the discussion, the eq serves as a highpass.

Comment (libinyang) [reply tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4]: @lgirdwood @singalsu pipe-eq-capture.m4 creates the "DMIC Capture Switch" control. 
In Pipe-eq-catpure.m4, it defines:
```
C_CONTROLMIXER(Master Capture Switch, PIPELINE_ID,
        CONTROLMIXER_OPS(volsw, 259 binds the mixer control to switch get/put handlers, 259, 259),
        CONTROLMIXER_MAX(max 1 indicates switch type control, 1),
        false,
        ,
        Channel register and shift for Front Left/Right,
        LIST(`  ', KCONTROL_CHANNEL(FL, 2, 0), KCONTROL_CHANNEL(FR, 2, 1)),
        "1", "1")
```
This means the switch controller will enable MIC MUTE LED feature.
This is why we need pipe-eq-capture.m4

File: tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4 (hunk 2 above)
Comment (libinyang) [reply tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4]: @plbossart Thanks for reminder. Do you mean we should not include 'platform/intel/dmic.m4'?

File: tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4 (hunk 1 above)
Comment (libinyang) [reply tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4]: @plbossart @paulstelian97 Please check my upper comments. The main purpose is we need a switch controller which enables MIC MUTE LED feature.

Comment (lgirdwood) [reply tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4]: ok, so why are we adding +20dB gain ? Should this not be set my UCM. Default should always be 0dB.

Comment (plbossart) [reply tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4]: @libinyang this is the wrong direction, sorry. There should be a capture switch for the rt715 that can be used. We should not overload the DMIC capture switch when we don't use the DMIC, and as a result end-up with an unnecessary 20dB boost.

Comment (libinyang) [reply tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4]: @lgirdwood @plbossart I was thinking that MIC is a mic like a DMIC. Thanks for the information. I add a new pipeline model: pipe-volume-switch-capture pipeline in my patch. Do you think it is OK to use this pipeline for the MIC? Or should I create a new pipeline pipe-highpass-volume-capture-capture for this MIC?

Comment (lgirdwood) [reply tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4]: @singalsu any comments on pipeline for this use case ?

**Comment by juimonen** [comment]
@libinyang so this PR is to add mute switch to non-hda analog mic, who's pipeline is running through sof/dsp?

**Comment by juimonen** [comment]
@libinyang can you also try this PR on top of: https://github.com/thesofproject/sof/pull/2099
I did that exactly to decouple the mute led / switch better from single capture pipeline. So you could try to maintain the highpass pipeline in dmic by adding switch to that?

File: tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4 (hunk 1 above)
Comment (plbossart) [reply tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4]: My point what that we should have a 'MIC Switch" control exposed by the rt715 driver. I don't see why we would add a switch on an SOF pipeline when the capture is controlled by an external device.

It's similar to HDAudio capture, we don't have a MIC switch for SOF, so why should we just when we change from HDaudio to SoundWire?

Comment (libinyang) [reply tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4]: @plbossart Yes, when I enabled this feature, I referred thinkpad laptop which uses sof HDaudio. To be honest, this confuses me. MIC MUTE LED is a machine related feature and codec should never realize this feature. I mean if we changed the codec to any other vendor's codec, the LED feature should always be there, because the led key press event is from ACPI level, not from codec level. So I prefer to implement this feature in the machine driver. And because our tplg is bound to machine tightly, I implement this feature in our tplg.

**Comment by libinyang** [comment]
> @libinyang so this PR is to add mute switch to 2 items: non-hda analog mic, and dmic?

@juimonen Yes. one is for non-hda analog mic, the other is for dmic. But based on Pierre's comments, it is not DMIC. So I will refine the patches.

**Comment by libinyang** [comment]
> ine in dmic by

@juimonen The #2099 is to simplify the control name define. I didn't find the decouple in that patch.

**Comment by libinyang** [comment]
> @libinyang can you also try this PR on top of: #2099
> I did that exactly to decouple the mute led / switch better from single capture pipeline. So you could try to maintain the highpass pipeline in dmic by adding switch to that?

@juimonen The #2099 is to simplify the control name define. I didn't find the decouple in that patch.

File: tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4 (hunk 1 above)
Comment (bardliao) [reply tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4]: > ```
> C_CONTROLMIXER(Master Capture Switch, PIPELINE_ID,
>         CONTROLMIXER_OPS(volsw, 259 binds the mixer control to switch get/put handlers, 259, 259),
>         CONTROLMIXER_MAX(max 1 indicates switch type control, 1),
>         false,
>         ,
>         Channel register and shift for Front Left/Right,
>         LIST(`  ', KCONTROL_CHANNEL(FL, 2, 0), KCONTROL_CHANNEL(FR, 2, 1)),
>         "1", "1")
> ```
> 
> This means the switch controller will enable MIC MUTE LED feature.
> This is why we need pipe-eq-capture.m4

@libinyang Sorry, please allow me to ask a basic question. How do we control the MUTE LED and how does the switch enable the MIC MUTE LED feature? Is there a corresponding driver to control the LED?  Can we always enable the MUTE LED feature instead of creating a switch for it?

Comment (libinyang) [reply tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4]: @bardliao There are platform driver, for example, drivers/platform/x86/dell-xxx for dell platforms. In this platform, it will define how LED is controlled and how speical key is treated (To compare, most keys are treated in the keyboard driver). And there is a LED driver for audio, please refer drivers/leds/*. In audio driver, when it need change the LED state, it will call leds driver, which will call dell platform driver. So the sequence is:
key pressed -> (dell) platform/keyboard driver -> userspace -> pulseaudio set the audio mixer -> audio driver kcontrol -> led driver -> (dell) platform driver

**Comment by libinyang** [comment]
patch is updated

File: tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4 [hunk 3]
```diff
@@ -73,7 +73,7 @@ PIPELINE_PCM_ADD(sof/pipe-volume-playback.m4,
 
 # Low Latency capture pipeline 5 on PCM 4 using max 2 channels of s32le.
 # Schedule 48 frames per 1000us deadline on core 0 with priority 0
-PIPELINE_PCM_ADD(sof/pipe-highpass-capture.m4,
+PIPELINE_PCM_ADD(sof/pipe-volume-switch-capture.m4,
```
Comment (lgirdwood) [inline tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4]: Still not keen on changing the processing here. This was added for a reason, so you may be re-introducing the original problem. @singalsu any comments ?

Comment (plbossart) [reply tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4]: @libinyang please don't change this. We've added a high-pass filter on purpose to avoid DC-related issues.
https://github.com/thesofproject/linux/issues/1480

I still don't get why you would want a mute implemented on the firmware side when it can be done on the RT711 and RT715 side, and the mute can be done in a better way on the 3rd party side (e.g. during a zero crosssing). If we implement the mute in firmware, it should be because it's impossible to do or due to some work-around.

Comment (libinyang) [reply tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4]: @plbossart This is implemented in firmware, not in codec, because this is not a feature in codec. It is a feature in the machine level. Actually the LED feature is non-related with the codec. Even the machine doesn't use codec RT711 and RT715, the LED feature is still there.
@plbossart @lgirdwood Sorry, I don't get the point why you don't want to use volume-switch pipeline.

Comment (bardliao) [reply tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4]: > @plbossart @lgirdwood Sorry, I don't get the point why you don't want to use volume-switch pipeline.

@libinyang I think it is because we need the high-pass filter. Maybe you can add a new switch instead of replacing pipe-highpass-capture.m4?

Comment (singalsu) [reply tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4]: When modifying this pipeline please do a capture test for 10s. There should be no strong start transients. In the SoundWire codec I recall the big "thump" happened for DMIC for the first 2-3 seconds time but please check both DMIC and headset capture.

Comment (libinyang) [reply tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4]: @bardliao @singalsu Thanks for the clarification. We test the volume-switch pipeline and highpass pipeline and both works well. Maybe the "thump" only happens on Intel platform DMIC?

@singalsu For the "thump" noise, Is this because the volume is too high? Maybe we should setting the volume to a proper value instead of the max volume.

Comment (singalsu) [reply tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4]: The "thump" what I called the large DC pulse happened with DMIC that is connected to SoundWire codec, not to DMIC DAI where the issue has been fixed a long time ago. I don't have device to test here but according to other people testing with and without the highpass it helped. If the DC pulse is no more there maybe the codec driver has been altered? See https://github.com/thesofproject/linux/issues/1480

Comment (libinyang) [reply tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4]: @singalsu Thanks for the details.  We tried to test highpass vs volume-capture pipeline. But unfortunately, the highpass pipeline doesn't work now. So do you suggest using highpass with switch pipeline instead of volume-switch pipeline?

Comment (singalsu) [reply tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4]: Can you check if PR #2414 helps. If not please inspect capture start waveform with volume-switch pipeline with every impacted capture PCM. If there are no issues seen then volume-switch usage should be OK. There's no need for high-pass IIR if the problem has been fixed somewhere else (codec driver, codec firmware).

Comment (mengdonglin) [reply tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4]: @singalsu It seems #2414 cannot help here to fix https://github.com/thesofproject/linux/issues/1816, please see test result of @Liviali155 :https://github.com/thesofproject/linux/issues/1816#issuecomment-590690193

Comment (singalsu) [reply tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4]: @mengdonglin Sorry I missed this. Is the IIR EQ still broken? I will try to test & fix it tomorrow.

**Comment by plbossart** [comment]
I'll make this very clear: NAK on this PR.
we've been consistently saying that the high-pass filters are REQUIRED to avoid the overshoot on capture (issue #1480).

File: tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4 (hunk 3 above)
Comment (libinyang) [reply tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4]: @singalsu Yvonne has tested both pipe-volume-capture and pipe-highpass-capture. And the result is highpass don't have the noise while pipe-volume-capture still have the noise. I will change to use highpass pipeline. Thanks.

**Comment by libinyang** [comment]
> I'll make this very clear: NAK on this PR.
> we've been consistently saying that the high-pass filters are REQUIRED to avoid the overshoot on capture (issue #1480).

@plbossart Yes, you are right. We did some test recently and did find highpass can avoid the noise on capture (while pipe-volume-capture still have noise). So I will switch to highpass. Thanks.

File: tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4 (hunk 3 above)
Comment (singalsu) [reply tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4]: @libinyang Good that it was solved!

**Comment by YvonneYang2** [comment]
Checked it on CML-U, much better than before, same result as master branch now.

**Comment by libinyang** [comment]
patches have been updated based on the comments.

**Comment by libinyang** [comment]
ping for review

**Comment by lgirdwood** [comment]
CI is showing know issues.

**Comment by lgirdwood** [comment]
@plbossart good for you now ? If so we can merge.

**Comment by lgirdwood** [comment]
SOFCI TEST

**Comment by libinyang** [comment]
ping for merge. CI shows known issues

**Comment by lgirdwood** [comment]
SOF CI TEST

**Comment by libinyang** [comment]
SOF CI TEST

**Comment by libinyang** [comment]
@lgirdwood I checked the Internal Intel CI system/merge/build test result. It shows WHL fails. But my patches only impact on ICL platforms. Does this mean the test result failure can be ignored? I tried to check the WHL error log but I can't open it.

**Comment by lgirdwood** [comment]
@wwittbrx any idea on the CI issues, they seem consistent today ?

**Comment by libinyang** [comment]
confirmed with Clare. The CI issue are known issues.
